### PR TITLE
docs(clouddriver): guide to configuring clouddriver with mysql

### PR DESCRIPTION
**Pull request by asher**

### CODE REVIEW

**Comment by asher** [comment]
The code referenced here is in clouddriver master but not yet in a versioned release. Would we normally wait for the next release (1.14) to land before merging related docs?

File: setup/productionize/persistence/clouddriver-sql.md
Comment (emjburns) [inline setup/productionize/persistence/clouddriver-sql.md]: Do you want to give any indication of what "small" and "large" are?

File: setup/productionize/persistence/clouddriver-sql.md
Comment (emjburns) [inline setup/productionize/persistence/clouddriver-sql.md:135]: Do you want to add a bit more about how to migrate? I don't think we have any docs in general on how to set up a database migration, and I think it would be a good conclusion to this guide.

**Comment by emjburns** [comment]
👍 This is great

File: setup/productionize/persistence/clouddriver-sql.md
Comment (asher) [reply setup/productionize/persistence/clouddriver-sql.md:135]: Good suggestion. There's no one way to do it, and I don't know how it would be done in a Halyard managed environment, so I think I'll write a "How Netflix Live Migrated Clouddriver from Redis to SQL" section.

File: setup/productionize/persistence/clouddriver-sql.md
Comment (dorbin) [inline setup/productionize/persistence/clouddriver-sql.md]:
```suggestion
You can configure Clouddriver to use a MySQL compatible database in place of Redis for all of its persistence use cases. These are:
```

File: setup/productionize/persistence/clouddriver-sql.md
Comment (dorbin) [inline setup/productionize/persistence/clouddriver-sql.md]:
```suggestion
You can also use Redis for one of the above, and SQL for another. A `DualTaskRepository` class is provided to enable online migrations between Redis and SQL in either direction without impact to running pipelines.
```

File: setup/productionize/persistence/clouddriver-sql.md
Comment (dorbin) [inline setup/productionize/persistence/clouddriver-sql.md]:
```suggestion
Growth in Netflix's infrastructure footprint over time exposed a number of scaling challenges with Clouddriver's original Redis-backed caching implementation. For example, the functional use of unordered Redis sets as secondary indexes incurs linear cost while also making it challenging to shard the keyspace without hotspots. Additionally, Clouddriver's data is highly relational, making it a good candidate for modeling to either relational or graph databases.
```

File: setup/productionize/persistence/clouddriver-sql.md
Comment (dorbin) [inline setup/productionize/persistence/clouddriver-sql.md]:
```suggestion
Small Spinnaker installations (managing up-to a few thousand instances across all applications) are unlikely to see significant performance or cost gains by running Clouddriver on an RDBMS as opposed to Redis. In the Netflix production environment, we have seen improvements of up to 400% in the time it takes to load the clusters view for large applications (hundreds of server groups with thousands of instances) since migrating from Redis to Aurora. We also see greater consistency in 99th percentile times under peak load and have reduced the number of instances required to operate Clouddriver. Besides improving performance at scale, we believe the SQL implementation is better suited to future growth. We've load tested Clouddriver backed by Aurora with the Netflix Production data set at more than 10x traffic rates, which previously resulted in Redis related outages.
```

File: setup/productionize/persistence/clouddriver-sql.md
Comment (dorbin) [inline setup/productionize/persistence/clouddriver-sql.md]: Regarding "...it is preferable...," what happens if the full dataset doesn't fit? 
Maybe just say, "make sure" rather than "it is preferable to ensure"?

File: setup/productionize/persistence/clouddriver-sql.md
Comment (dorbin) [inline setup/productionize/persistence/clouddriver-sql.md]:
```suggestion
Throughput and write latency are important considerations, especially for Spinnaker installations that manage many accounts and regions. Each distinct `(cloudProvider, account, region, resourceType)` tuple generally gets an independently scheduled caching-agent instance within Clouddriver. The number of these instances per Clouddriver configuration can be considered the upper bound for concurrent write-heavy sessions to the database, if not limited by the `sql.agent.maxConcurrentAgents` property. Caching-agent write throughput can directly impact deployment times, especially in large environments with many accounts or multiple cloud providers. Clouddriver only tries to write new or modified resources to the database however, moderating write requirements after initial population.
```

File: setup/productionize/persistence/clouddriver-sql.md
Comment (dorbin) [inline setup/productionize/persistence/clouddriver-sql.md]:
```suggestion
Before you deploy Clouddriver, you need to manually create a logical database and user grants.
```

File: setup/productionize/persistence/clouddriver-sql.md
Comment (dorbin) [inline setup/productionize/persistence/clouddriver-sql.md]: "should be considered"...

"are recommended"? Or required?

Along those lines, if you can, remove the instances of "should" in the list of parameters.  Or make it clear what the consquences are of using something different for each one.

File: setup/productionize/persistence/clouddriver-sql.md
Comment (dorbin) [inline setup/productionize/persistence/clouddriver-sql.md]:
```suggestion
The following yaml-based parameters provide a Clouddriver configuration that entirely uses MySQL in place of Redis. [Halyard](/reference/halyard/) users can provide these as overrides via a `clouddriver-local.yml` file. Halyard does not yet natively support Clouddriver SQL configuration.
```

Comment (dorbin) [inline setup/productionize/persistence/clouddriver-sql.md]: Could you swap the 2nd and 3rd sentences?

File: setup/productionize/persistence/clouddriver-sql.md
Comment (dorbin) [inline setup/productionize/persistence/clouddriver-sql.md]:
```suggestion
## Configure Clouddriver to use MySQL
```

File: setup/productionize/persistence/clouddriver-sql.md
Comment (dorbin) [inline setup/productionize/persistence/clouddriver-sql.md]:
```suggestion
The above yaml configures Clouddriver to use a MySQL table as part of a locking service for the caching agent scheduler. This works well for Netflix in production against Aurora, but could result in poor overall performance for all of Clouddriver, depending on how your database is configured. If you observe high database CPU utilization, lock contention, or if caching agents aren't consistently running at the expected frequency, you may want to try using the Redis-based scheduler to determine if this is a contributing factor. 

The following modification to the above example configures Clouddriver to use SQL for the cache and task repository, and Redis for agent scheduling.
```

File: setup/productionize/persistence/clouddriver-sql.md
Comment (dorbin) [inline setup/productionize/persistence/clouddriver-sql.md]:
```suggestion
If you're migrating from Redis to SQL in a production environment, where you need to avoid pipeline failures and downtime, you can modify the configuration, as shown below, to configure Clouddriver use SQL for caching and new tasks, but with fallback reads to Redis for tasks not found in the SQL database.
```

**Review by dorbin — COMMENTED** [review]
Mostly just some wording edits/suggestions, @asher .

Thanks for the doc!

**Comment by asher** [comment]
@dorbin great suggestions, adopted them all, thanks! I added a "How Netflix Migrated Clouddriver from Redis to SQL" section at the bottom at @emjburns suggestion if you'd like to give it read.

File: setup/productionize/persistence/clouddriver-sql.md
Comment (emjburns) [inline setup/productionize/persistence/clouddriver-sql.md:237]: This section is great! Thank you @asher

File: setup/productionize/persistence/clouddriver-sql.md
Comment (dorbin) [inline setup/productionize/persistence/clouddriver-sql.md:162]: @asher 
Might be helpful to identify which config file this is.

**Review by dorbin — APPROVED** [review]

**Pull request merged**
